### PR TITLE
[Feature]Support schema changes in Iceberg catalog

### PR DESCRIPTION
**Pull request by Zhangg7723**
Why I'm doing:

> For better Iceberg format support

What I'm doing:

> Provide new feature for iceberg tables, include alter table with add/drop/modify/rename column.

Fixes #issue

> This PR will fix these errors
> https://github.com/StarRocks/starrocks/issues/30904
> https://github.com/StarRocks/starrocks/issues/30905

SQL demo supported:
```
Alter table <tbl_name> add column col1 string comment 'col1';
Alter table <tbl_name> add column col2 string null comment 'col2' after id;
Alter table <tbl_name> add column (col3 string comment 'col3', col4 INT);
Alter table <tbl_name> rename column col1 to col11;
Alter table <tbl_name> modify column col4 BIGINT;
Alter table <tbl_name> drop column col4;
```

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/DDLStmtExecutor.java [hunk 1]
```diff
@@ -335,9 +336,16 @@ public ShowResultSet visitCancelRefreshMaterializedViewStatement(CancelRefreshMa
...
-                context.getGlobalStateMgr().alterTable(stmt);
-            });
+            String catalogName = stmt.getCatalogName();
+            if (CatalogMgr.isInternalCatalog(catalogName)) {
+                ErrorReport.wrapWithRuntimeException(() -> {
+                    context.getGlobalStateMgr().alterTable(stmt);
+                });
+            } else {
+                ErrorReport.wrapWithRuntimeException(() -> {
+                    context.getGlobalStateMgr().getMetadataMgr().alterTable(stmt);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/qe/DDLStmtExecutor.java:346]: both of two catalogs can use `context.getGlobalStateMgr().getMetadataMgr().alterTable(stmt);`

File: fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java [hunk 2]
```diff
@@ -259,6 +261,27 @@ public boolean createTable(CreateTableStmt stmt) throws DdlException {
...
+        Optional<ConnectorMetadata> connectorMetadata = getOptionalMetadata(catalogName);
+
+        if (connectorMetadata.isPresent()) {
+            String dbName = stmt.getDbName();
+            String tableName = stmt.getTableName();
+            if (getDb(catalogName, dbName) == null) {
+                throw new DdlException("Database '" + dbName + "' does not exist in catalog '" + catalogName + "'");
+            }
+
+            if (!listTableNames(catalogName, dbName).contains(tableName)) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java:275]: nit: after we implement the tableExists() interface. you can use it to check.

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java [hunk 3]
```diff
@@ -203,6 +214,65 @@ public boolean createTable(CreateTableStmt stmt) throws DdlException {
...
+                throw new StarRocksConnectorException(
+                        "Unsupported alter operation for iceberg connector: " + clause.toString());
+            }
+        }
+
+        commitAlterTable(table, schemaChanges, tableChanges);
+
+        synchronized(this) {
+            tables.remove(TableIdentifier.of(dbName, tableName));
+            icebergCatalog.refreshTable(dbName, tableName, jobPlanningExecutor);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java:253]: catch refresh table exception and then invalidate the cache of cachingIcebergCatalog.

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergApiConverter.java [hunk 4]
```diff
@@ -60,6 +71,7 @@ public class IcebergApiConverter {
     private static final Logger LOG = LogManager.getLogger(IcebergApiConverter.class);
     public static final String PARTITION_NULL_VALUE = "null";
     private static final int FAKE_FIELD_ID = -1;
+    private static final String ALLOW_INCOMPATIBLE_CHANGES = "allow_incompatible_changes";
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergApiConverter.java:74]: what's this? why use this properties?

**Comment by stephen-shelby** [comment]
it's best to add some sql_test.

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergApiConverter.java (hunk 4 above)
Comment (Zhangg7723) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergApiConverter.java:74]: This option allows incompatible changes to be made to a schema, For example, adding a required column and attempting to read old data files without that column will cause a failure in iceberg table.
We can set this option in alter column statement like:
```
alter table ice_table02 
add column col10 INT not null default '0'
PROPERTIES
("allow_incompatible_changes"="true")
```

**Comment by Zhangg7723** [comment]
> it's best to add some sql_test.

**Comment by Zhangg7723** [comment]
> it's best to add some sql_test.

Thanks for your review, I will add sql test later.

File: fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java (hunk 2 above)
Comment (Zhangg7723) [reply fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java:275]: nice

**Comment by Zhangg7723** [comment]
@stephen-shelby hi, please review again, I added sql test and updated code.

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergApiConverter.java [hunk 5]
```diff
@@ -290,4 +303,98 @@ public static Map<String, String> rebuildCreateTableProperties(Map<String, Strin
...
+                }
+                updateSchema.updateColumn(column.getName(), colType.asPrimitiveType());
+
+                // UPDATE comment
+                if (column.getComment() != null) {
+                    updateSchema.updateColumnDoc(column.getName(), column.getComment());
+                }
+
+                // NOT NULL / NULL
+                if (column.isAllowNull()) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergApiConverter.java:376]: Column must be nullable in the starrocks external table now.

File: fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java (hunk 2 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java:275]: you could use tableExists now

**Review by stephen-shelby — DISMISSED** [review]

**Review by stephen-shelby — DISMISSED** [review]

**Review by Youngwb — DISMISSED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**